ls1intum/Artemis
# `Tutorial groups`: Give students information about the current utilization and average attendance

### Checklist
#### General
#### Server
#### Client

### Motivation and Context
- We want to show the average attendance and utilization of a tutorial group, so students know which are crowded groups and which are empty groups where a more personal tutoring is possible.

### Description
- Added the `averageAttendance` transient property to the `tutorialGroup` entiy. It is calculated in the following way:
```
   /**
     * Sets the averageAttendance transient field of the given tutorial group
     * <p>
     * Calculation:
     * <ul>
     *     <li>Get set of the last three completed sessions (or less than three if not more available) </li>
     *     <li>Remove sessions without attendance data (null) from the set</li>
     *     <li>If set is empty, set attendance average of tutorial group to null (meaning could not be determined)</li>
     *     <li>If set is non empty, set the attendance average of the tutorial group to the arithmetic mean</li>
     * </ul>
     *
     * @param tutorialGroup the tutorial group to set the averageAttendance for
     */
```

Now on the client I have added two new columns to the tutorial group table. One for the `averageAttendance` and one for the `utilization` which is calculated by `averageAttendance/capacity`. The `utilization` is displayed via a vertical progress bar. Help texts explain the calcultion via tooltips.

### Steps for Testing
- Create a tutorial group with at least three sessions in the past. (with capacity set)
- Add attendance data to these sessions.
- Check that the average attendance and utilization is correctly calculated.



### Review Progress
#### Code Review
#### Manual Tests

### Test Coverage
- As you can see from the code I have added both server and client tests to cover the changes.

| Component                                         | Statement Coverage | Branch Coverage |
|---------------------------------------------------|--------------------|-----------------|
| vertical-progress-bar.component.ts                | 97.56%             | 90.47%          |
| tutorial-group-utilization-indicator.component.ts | 100%               | 100%            |
| TutorialGroupService.java                         | 94%                | 74%             |


### Screenshots
<img width="1073" alt="image" src="https://user-images.githubusercontent.com/29718932/218308870-d60f61a2-e92a-4d55-aba8-5926caed40c6.png">
<img width="534" alt="image" src="https://user-images.githubusercontent.com/29718932/218309079-3ab6533f-0107-4eee-88ff-f2a2deee302b.png">
<img width="1435" alt="image" src="https://user-images.githubusercontent.com/29718932/218313490-6f23ab29-6316-45e9-839b-505e4d2703f0.png">
